fl4health/clients/fed_pca_client.py

```diff
@@ -0,0 +1,144 @@
+import random
+import string
+from logging import INFO
+from pathlib import Path
+from typing import Dict, Tuple, Type, TypeVar
+
+import torch
+from flwr.client.numpy_client import NumPyClient
+from flwr.common import Config, NDArrays, Scalar
+from flwr.common.logger import log
+from torch import Tensor
+from torch.utils.data import DataLoader
+
+from fl4health.model_bases.pca import PCAModule
+from fl4health.parameter_exchange.full_exchanger import FullParameterExchanger
+
+T = TypeVar("T")
+
+
+class FedPCAClient(NumPyClient):
+    def __init__(self, data_path: Path, device: torch.device, model_save_path: Path) -> None:
+        """
+        Client that facilitates the execution of federated PCA.
+        """
+        self.client_name = self.generate_hash()
+        self.model: PCAModule
+        self.initialized = False
+        self.data_path = data_path
+        self.model_save_path = model_save_path
+        self.device = device
+        self.train_loader: DataLoader
+        self.val_loader: DataLoader
+        self.num_train_samples: int
+        self.num_val_samples: int
+        self.parameter_exchanger: FullParameterExchanger
+
+    def generate_hash(self, length: int = 8) -> str:
+        return "".join(random.choice(string.ascii_lowercase) for _ in range(length))
+
+    def get_parameters(self, config: Config) -> NDArrays:
+        assert self.model is not None and self.parameter_exchanger is not None
+        return self.parameter_exchanger.push_parameters(self.model, config=config)
+
+    def set_parameters(self, parameters: NDArrays, config: Config) -> None:
+        self.parameter_exchanger.pull_parameters(parameters, self.model, config)
```

Comment:
This is a bit different than the standard flow, as we're saving the model on setting the parameters. A comment would be useful to explain why.

Reply:
Yes for sure. Since federated PCA only runs for one round, the parameters obtained here are the merged principal components, and because currently for downstream tasks we need to save these parameters, I thought this is the most appropriate place to save them. I've added a comment to this method as well.